server/mediaserver.go

```diff
@@ -540,6 +540,8 @@ func (s *LivepeerServer) registerConnection(rtmpStrm stream.RTMPVideoStream) (*r
 	if exists {
 		// We can only have one concurrent stream per ManifestID
 		s.connectionLock.Unlock()
+		cxn.sessManager.cleanup()
```

Comment:
@darkdarkdragon After doing the cleanup changes here - I noticed that instead of getting a 503 that I was getting before - I am now again getting 500s due to a new reason.

Closing the session manager created by the *second* segment, closes the B session for the old connection too for some reason. I don't understand why exactly this is happening, but I'll do a quick change without this cleanup call here and see if it works out.

Reply:
I now get
```
E1205 00:32:59.274014 1204668 mediaserver.go:857] http push error processing segment url=http://127.0.0.1:8936/live/test/1.ts manifestID=test err=Session ended
```
And sending any segment again on the same streamID too I keep getting this. So I think cleaning up the newly created session here is somehow closing both the old and new sessions.

Reply:
@jailuthra I believe the cause of the "Session ended" errors described in your [comment](https://github.com/livepeer/go-livepeer/pull/1672#issuecomment-739001491) were caused by calling `cxn.pl.Cleanup()` and not by calling `cxn.sessManager.cleanup()` (I'm assuming you were also calling `cxn.pl.Cleanup()` in a previous commit).

When two concurrent segments are received, the second call to `createRTMPStreamIDHandler()` will call [os.NewSession](https://github.com/livepeer/go-livepeer/blob/master/server/mediaserver.go#L276) with the same webhook manifestID as the first call to `createRTMPStreamIDHandler()`. In this case, `os.NewSession()` will return an already [cached storage session](https://github.com/livepeer/go-livepeer/blob/master/drivers/local.go#L44) since the first call to `createRTMPStreamIDHandler()` had created that storage session for the webhook manifestID. As a result when `cxn.pl.Cleanup()` is called in the code block that is executed if an RTMP conn already exists, the storage session for both the first and second segment will be [ended](https://github.com/livepeer/go-livepeer/blob/7138591ec3ee61d49afe1f4fa3720539f30c72ae/core/playlistmanager.go#L234) since the playlist for both RTMP conns reference the same storage session. Then, the next time that storage session is used to save data i.e. during the segment workflow, we get the "Session ended" error.

An easy way to avoid this error is to not call `cxn.pl.Cleanup()` in that RTMP conn already exists code block which is safe to do right now since the cleanup function is not responsible for anything else besides ending the storage session. But, if that cleanup function ever included anything besides ending the storage session then we'd after to reconsider omitting the `cxn.pl.Cleanup()` call...I suppose that might be preferable to increasing the time under lock (as is done in the latest commit) at least until we can figure out additional improvements to the code flow here...

Reply:
@yondonfu Oh I missed this comment in my notifications while sending the update. Thanks! will fix in a new update rn.